textmetric/metrics.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2021 Dmitry Vilensky-Pasechnyuk
+
+from collections import defaultdict
+import numpy as np
+
+from nltk.translate.meteor_score import single_meteor_score
+from nltk.translate.bleu_score import corpus_bleu, sentence_bleu
+from nltk.translate.chrf_score import corpus_chrf, sentence_chrf
+from rouge import Rouge
+from chrFpp import computeChrF
+from bert_score import score as bert_score
+
+from time import perf_counter
+
+
+class Metrics:
+    def __init__(self, hyps, refs):
+        self.hyps = hyps
+        self.refs = refs
+
+    def get_statistics(self, with_symbolic=False, bert=False):
+        result = {
+            'scores': {},
+            'score': {}
+        }
+        for func in dir(Metrics):
+            if callable(getattr(Metrics, func)) and func[0] != '_' and func != 'get_statistics':
+                if (not with_symbolic and 'symb' in func) or (not bert and func == 'bert'):
+                    continue
+                print(f'Calculating {func} metric')
+                start = perf_counter()
+                metric_result = getattr(Metrics, func)(self)
+                result['scores'].update(metric_result['scores'])
+                result['score'].update(metric_result['score'])
+                finish = perf_counter()
+                print('Metric is calculated. Elapsed time:', finish - start)
+
+        return result
+
+    def bert(self):
+        # TODO: add a proper scaling
+        (P, R, F), hashname = bert_score(
+            self.hyps, self.refs, lang="en", return_hash=True)
+        
+        return {
+            'scores': {
+                'bert-P': P.cpu().detach().numpy(),
+                'bert-R': R.cpu().detach().numpy(),
+                'bert-F': F.cpu().detach().numpy()
+            },
+            'score': {
+                'bert-P': P.mean().item(),
+                'bert-R': R.mean().item(),
+                'bert-F': F.mean().item()
+            }
+        }
+
+    def meteor(self):
+        scores = []
+        for hyp, ref in zip(self.hyps, self.refs):
+            scores.append(single_meteor_score(ref, hyp))
+
+        return {
+            'scores': {
+                'meteor': np.array(scores)
+            },
+            'score': {
+                'meteor': np.mean(scores)
+            }
+        }
+
+    def bleu(self):
+
+        # TODO: there are in fact the standard weights of BLEU, isn't it?
+        def bleu_weights(l):
+            length = len(l)
+            if length >= 4:
+                return (0.25, 0.25, 0.25, 0.25)
+            wt = [1 / length for _ in range(length)]
```

Comment:
Here we should probably do 
```suggestion
            wt = [1 / (length-1) for _ in range(length-1)]
```
with [1, 0, 0, 0] weights for 1-token case. This is because our current configuration punishes short-token examples too harshly.
Another option is to use some smoothing function of nltk: https://www.nltk.org/_modules/nltk/translate/bleu_score.html (method 5 is the best according to https://sarahfakhoury.com/2021-FSE-Summarization-Metrics.pdf); the weights are then standard [0.25, 0.25, 0.25, 0.25]

All this stuff is a huge dumpster fire mess. I'd suggest computing two different versions of BLEU (one I suggest and one with a smoothing method and use the better one in the paper.

Reply:
Fixed